sql/analyzer/resolve_columns.go

```diff
@@ -38,6 +38,41 @@ func checkAliases(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error) {
 	return n, err
 }
 
+func checkNoTuplesProjected(ctx *sql.Context, a *Analyzer, node sql.Node) (sql.Node, error) {
+	span, _ := ctx.Span("no_tuples_projected")
+	defer span.Finish()
+
+	a.Log("check no tuples as in projection, node of type: %T", node)
+	return node.TransformUp(func(node sql.Node) (sql.Node, error) {
+		project, ok := node.(*plan.Project)
```

Comment:
Project and GroupBy share a common interface: sql.Expressioner, which returns all expressions in a node.
You could leverage that common interface to reduce the code of this function to the following:

```go
switch node := node.(type) {
case *plan.Project, *plan.GroupBy:
        for _, e := range node.(sql.Expressioner).Expressions() {
                if sql.IsTuple(e.Type()) {
                        return nil, ErrTupleProjected.New()
                }
        }
}
```

Note that in this snippet I added a fix to your code. Not always you will have an `expression.Tuple` at the top level of an expression. The safe way to check if an expression is a tuple is if its type is a tuple. For this, we have a helper function named `sql.IsTuple`, so you can do `sql.IsTuple(expr.Type())`.

Reply:
Done (in the existing rule that wasn't doing it that way).